packages/@orbit/jsonapi-websocket/src/lib/connection.ts

```diff
@@ -0,0 +1,193 @@
+import { evented, Evented, Listener } from '@orbit/core';
+
+import ConnectionMonitor, { PollInterval } from './connection-monitor';
+
+export interface ConnectionSettings {
+  protocols: string[];
+  reopenDelay?: number;
+  pollInterval?: PollInterval;
+  staleThreshold?: number;
+}
+
+@evented
+export default class Connection implements Evented {
```

Comment:
how is this going to work for other websocket implementations? 

like signalr, for example, has a protocol that allows switching between long polling, polling, and websockets (and maybe one other one?).

it would be cool if there was a way to have an adapter layer so that someone could plugin in their own existing websocket implementation and just have it hook in to orbit

Reply:
I fully agree with you. this is a first pass. But I definitely would like to have an adapter layer. If you have a suggestion for an interface I would love to hear it :)

Reply:
I think it could be a pretty lightweight interface:

```ts
sendToServer(operationsDocument: OperationsDocument)?: OperationsDocument {
 // do signal-r hookup stuff in here  
}
handleMessageFromServer(operationsDocument: OperationsDocument): void {
 // do signal-r hookup stuff in here
}
// maybe something like that?
```

Reply:
After some thoughts, and given what we landed in `jsonapi-serializer`, I think trying to build a general socket source would be premature. Just as we don't have a general network source. I would really like in the future to have more generic sources, but for now I think the most important part is we are shipping primitives that allow for easy experiments. For example, this is a simple `ActionCable` source: https://gist.github.com/tchak/9dfa1e8b17946e985aaec1480dddde5e. As you can see, it's basically one method source.